In [1]:
%matplotlib inline
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns
from sklearn.linear_model import LogisticRegression
from sklearn.model_selection import train_test_split
from sklearn.preprocessing import PolynomialFeatures
from sklearn.datasets import load_digits


In [2]:
x,y = load_digits(return_X_y=True)
x,y

(array([[ 0.,  0.,  5., ...,  0.,  0.,  0.],
        [ 0.,  0.,  0., ..., 10.,  0.,  0.],
        [ 0.,  0.,  0., ..., 16.,  9.,  0.],
        ...,
        [ 0.,  0.,  1., ...,  6.,  0.,  0.],
        [ 0.,  0.,  2., ..., 12.,  0.,  0.],
        [ 0.,  0., 10., ..., 12.,  1.,  0.]], shape=(1797, 64)),
 array([0, 1, 2, ..., 8, 9, 8], shape=(1797,)))

In [3]:
X_train, X_test, y_train, y_test = train_test_split(x,y,test_size=0.2)

In [4]:
def train(X_train, X_test, y_train, y_test,max_iter=10000 , C=1 ,solver='lbfgs',dual=False):
    model = LogisticRegression(max_iter=max_iter , C=C ,solver=solver,dual=dual)
    model.fit(X_train,y_train)
    s_train = model.score(X_train,y_train)
    s_test = model.score(X_test,y_test)
    return model , f'score_train :{s_train} , score_test : {s_test}'

In [ ]:
solver = ["lbfgs", "liblinear", "newton-cg", "newton-cholesky", "sag", "saga"]  
c=[100,10,2,1,0.8,0.5,0.1]
for i_solver in range(len(solver)):
    for i_C in range(len(c)):
        print(f'{solver[i_solver]} -> {train(X_train, X_test, y_train, y_test, solver=solver[i_solver] ,C=c[i_C])}')

lbfgs -> (LogisticRegression(C=100, max_iter=10000), 'score_train :1.0 , score_test : 0.9638888888888889')
lbfgs -> (LogisticRegression(C=10, max_iter=10000), 'score_train :1.0 , score_test : 0.9583333333333334')
lbfgs -> (LogisticRegression(C=2, max_iter=10000), 'score_train :1.0 , score_test : 0.9583333333333334')
lbfgs -> (LogisticRegression(C=1, max_iter=10000), 'score_train :1.0 , score_test : 0.9611111111111111')
lbfgs -> (LogisticRegression(C=0.8, max_iter=10000), 'score_train :1.0 , score_test : 0.9611111111111111')
lbfgs -> (LogisticRegression(C=0.5, max_iter=10000), 'score_train :1.0 , score_test : 0.9611111111111111')
lbfgs -> (LogisticRegression(C=0.1, max_iter=10000), 'score_train :0.9993041057759221 , score_test : 0.9583333333333334')
liblinear -> (LogisticRegression(C=100, max_iter=10000, solver='liblinear'), 'score_train :0.9986082115518441 , score_test : 0.9305555555555556')
liblinear -> (LogisticRegression(C=10, max_iter=10000, solver='liblinear'), 'score_train :0.997

- lbfgs -> (LogisticRegression(C=0.1, max_iter=10000), 'score_train :0.9986082115518441 , score_test : 0.9805555555555555')
- liblinear -> (LogisticRegression(C=10, max_iter=10000, solver='liblinear'), 'score_train :0.9972164231036882 , score_test : 0.9611111111111111')
- liblinear -> (LogisticRegression(C=2, max_iter=10000, solver='liblinear'), 'score_train :0.9958246346555324 , score_test : 0.975')
- liblinear -> (LogisticRegression(C=1, max_iter=10000, solver='liblinear'), 'score_train :0.9951287404314544 , score_test : 0.975')
- newton-cg -> (LogisticRegression(C=0.1, max_iter=10000, solver='newton-cg'), 'score_train :0.9986082115518441 , score_test : 0.9805555555555555')
- sag -> (LogisticRegression(C=0.1, max_iter=10000, solver='sag'), 'score_train :0.9986082115518441 , score_test : 0.9805555555555555')
- saga -> (LogisticRegression(C=0.1, max_iter=10000, solver='saga'), 'score_train :0.9986082115518441 , score_test : 0.9805555555555555')

In [ ]:
solver = ["liblinear"]  
c=[100,10,2,1,0.8,0.5,0.1]
for i_solver in range(len(solver)):
    for i_C in range(len(c)):
        print(f'{solver[i_solver]} -> {train(X_train, X_test, y_train, y_test, solver=solver[i_solver] ,C=c[i_C],dual=True ,max_iter=10000 )}')

c:\Users\Taha\Desktop\python_p\AI\Learning_venv\Lib\site-packages\sklearn\svm\_base.py:1249: ConvergenceWarning: Liblinear failed to converge, increase the number of iterations.
  warnings.warn(


liblinear -> (LogisticRegression(C=100, dual=True, max_iter=100000, solver='liblinear'), 'score_train :0.9986082115518441 , score_test : 0.9583333333333334')


In [26]:
train(X_train, X_test, y_train, y_test, solver='liblinear',max_iter=10000 ,C=1.1)

(LogisticRegression(C=1.1, max_iter=10000, solver='liblinear'),
 'score_train :0.9958246346555324 , score_test : 0.9527777777777777')